### ISSUE TITLE
adding shared field to link related logs

### ISSUE THREAD

**Issue by crollorc** [logstash-plugins/logstash-filter-aggregate#52]
I have a query on the logstash discussion board which is related to the aggregate filter:
	https://discuss.elastic.co/t/logstash-adding-shared-field-to-link-related-logs/68933/1

As I noted in my post I haven't found a way to use the aggregate filter to create a shared event field which links related logs due to the master log arriving after the other logs.

The config I used is like the following repeated for call_ids[0] -> call_ids [4]:
      
```
if ([call_ids][0] != "") {
                aggregate {
                        task_id         => "%{call_ids[0]}"
                        code            => "
                                        if !map.key?(:call_id_link)
                                                if event.get('call_id_link')
                                                        map['call_id_link'] = event.get('call_id_link')
                                                else
                                                        map['call_id_link'] = event.get('[@metadata][fingerprint]')
                                                        event.set('call_id_link', map['call_id_link'])
                                                end
                                        end
                                        "
                        timeout         => 60
                }
        }

```

do you have any advise on how to proceed?

thanks for your help,

colm

**Comment by fbaligand** [comment]
Could you provide some input json documents before aggregate plugin, and the expected aggregated event after aggregate plugin ?

**Comment by crollorc** [comment]
Hi Fabien,

Thanks for getting back to me.

For an example SIP call, multiple devices will generate call logs which will be loaded within a couple of seconds of each other from files. 

These logs will be groked and amongst other things, call_ids will be extracted. 

Here are some sample logs and the associated call_ids:
1) MGC#1
> **log:** RP$25d T0db 1 0 20161214 082234 4 9 10 ac1a8015 0 0 1 00A 4 390694800319 4 390694800319 0 1 4 393484959705 001 0 4 393484959705 1 8 F10 FROM_GCS_UK2 314702-249169 loop[IC_UK_VSIP38362_OMEGA_1] loop[gw_sip,IC_UK_VSIP37927_BICS_2] 0 D00d 14894 50f0c6 E007 020100 H01b 4 390694800319 10 SERVNODE S015 1 1 4 393484959705 0 G033 20161214 08223402 0000000183 0000000693 0000001065 V013 8 20 0 0 0 SIP_OUT V012 8 20 0 0 0 SIP_IN Z02e **27047-VV-06dfc1f0-56ed86442** V013 8 20 0 0 0 SIP_OUT Z02e **28939-FV-06dfc1ee-3fe84a435** Z035 **7c58b6906383931c66a7d8d2530c6b3c**
> **call_ids:** **27047-VV-06dfc1f0-56ed86442, 28939-FV-06dfc1ee-3fe84a435, 7c58b6906383931c66a7d8d2530c6b3c**

2) SBC#1
> **log:** I2$190 T0b4 0 1 20161214 082233 4 4 11 ac1a800b 0 0 0 000 4 390694800319 0 X 0 0 4 393484959705 000 0 0 X 0 0 F10 X X IC_UK_VSIP38362_OMEGA_1 172.26.160.20:5060 0 Z035 **7c58b6906383931c66a7d8d2530c6b3c** X028 "IBCF-2-148170375385588622905679" X X X G02a 20161214 08223385 ? 0000000710 0000001080 O036 11000 60800 16640 380 104 0 0 0 X 212.126.62.75:34156
> **call_ids:** **7c58b6906383931c66a7d8d2530c6b3c**

3) SBC#2
> **log:** I2$1ad T0d8 0 0 20161214 082234 4 4 10 ac1a800b 0 0 0 000 4 390694800319 4 390694800319 0 0 4 393484959705 000 0 0 X 0 0 F10 X X 172.26.160.20:5060 IC_UK_VSIP37927_BICS_2 0 Z02e **27047-VV-06dfc1f0-56ed86442** X028 "IBCF-2-148170375412027222905684" X X X G02a 20161214 08223412 ? 0000000683 0000001056 O036 10000 16640 61120 104 382 0 0 0 X 212.126.62.77:34864
> **call_ids:** **27047-VV-06dfc1f0-56ed86442**

As above, the MGC type logs are like a master log and will always contain the call_ids of all associated logs.

I would like to add a shared unique id (call_id_link) to each of the events which have related call_ids. In tests, I've used the output of the fingerprint filter for this purpurse.
1) MGC#1
> **call_link_id:** 63420e94e87d9a17f65162e83e948468cd419c52
2) SBC#1
> **call_link_id:** 63420e94e87d9a17f65162e83e948468cd419c52
3) SBC#2
> **call_link_id:** 63420e94e87d9a17f65162e83e948468cd419c52
	
In my example above where the MGC logs arrives first, I can get separate aggregate filters for each call_id to work.

However, in most cases, the MGC log does not arrive first and I do not know how to get the aggregate filters to work in that case.

Apologies for wasting your time if the aggregate filter is simply in-appropriate for this use case.

Thanks for your help,

Colm

**Comment by crollorc** [comment]
Hi Fabien,

Did you get a chance to look at my post and do you have any advice on how to proceed?

Thanks again for any help.

**Comment by fbaligand** [comment]
Hi @crollorc 

First, sorry for my late answer.
Then, if I understand well, when you get a MGC log, you generate a `call_link_id` field. And when you get a SBC log, you want to retrieve `call_link_id` field to copy it into SBC log.
Is this your need ?

But your problem is that MGC log always come after SBC logs.

If your MGC log always arrive after all SBC related logs, I have an idea to process your case.
The idea is to keep SBC logs inside logstash until MGC log arrives.
```
filter  {
	
	if [type] == 'SBC' {
		aggregate {
			task_id => "%{SBC_call_id}"
			code => "map.merge!(event.to_hash)"
		}
		drop {}
	}

	if [type] == 'MGC' {
		# TODO : set call_link_id in MGC log ...

		clone {
			clones => ["merge_sbc_log"]
		}
		if [type] == "merge_sbc_log" {
			split {
				field => "call_ids"
				target => "call_id"
			}
			aggregate {
				task_id => "%{call_id}"
				code => "
					map['call_link_id'] = event.get('call_link_id')
					event.overwrite(LogStash::Event.new(map))
				"
				end_of_task => true
			}
		}
	}
}
```

What do you think about ?

**Comment by crollorc** [comment]
Hi Fabien,

That is extremely helpful as I haven't used the clone or split filters before and have no ruby expertise.

> If your MGC log always arrives after all SBC related logs, I have an idea to process your case.

Unfortunately, the MGC log usually arrives in the middle and can really come anywhere in the sequence.

In any case, I tried the following to see if at least the first SBC log could be linked to the MGC log (I changed the SBC code a little to try to permit later SBC logs through):

```
        if ([call_ids][0] != "") {
                if ([type] == "sbc") {
                        aggregate {
                                task_id                 => "%{call_ids[0]}"
                                code                    => "
                                                                if !map.key?(:call_id_link)
                                                                        map.merge!(event.to_hash)
                                                                        event.cancel
                                                                end
                                                        "
                                timeout                 => 60
                                add_tag                 => [ "call_ids_0"]
                        }
                }
                if ([type] == 'mgc') {
                        ruby {
                                code                    => "
                                                                event.set('call_id_link', event.get('[@metadata][fingerprint]'))
                                                        "
                        }
                        clone {
                                clones                  => ["merge_sbc_log"]
                        }
                        if ([type] == "merge_sbc_log") {
                                split {
                                        field           => "call_ids"
                                        target          => "call_id"
                                }
                                aggregate {
                                        task_id         => "%{call_id}"
                                        code            => "
                                                                map['call_link_id'] = event.get('call_id_link')
                                                                event.overwrite(LogStash::Event.new(map))
                                                        "
                                        end_of_task     => true
                                        add_tag         => [ "call_ids_0"]
                                }
                        }
                }
        }


```		

Unfortunately, this replaces all the SBC logs with a single entry like this:

```
{
  "_index": "logstash-%{type}-2016.12.19",
  "_type": "logs",
  "_id": "%{[@metadata][fingerprint]}",
  "_score": null,
  "_source": {
    "@timestamp": "2016-12-19T12:22:39.962Z",
    "@version": "1",
    "call_link_id": null,
    "tags": [
      "call_ids_0"
    ]
  },
  "fields": {
    "@timestamp": [
      1482150159962
    ]
  },
  "sort": [
    1482150159962
  ]
}

```

Thanks again for your help though!

**Comment by fbaligand** [comment]
Question : in the following block code, why do you add a `if` statement ?
```
if !map.key?(:call_id_link)
    map.merge!(event.to_hash)
    event.cancel
end
```

**Comment by crollorc** [comment]
@fbaligand 

> Question : in the following block code, why do you add a if statement ?

The most frequent sequence of logs is SBC, MGC, SBC and (it may be a mistake) but hoped this code would prevent the second SBC log being dropped.

**Comment by fbaligand** [comment]
But if I understood well, each sbc log has its own call id. 
So, given that task_id is "call_ids[0]", there is no risk that 2 sbc logs share the same aggregate map. No ?

**Comment by crollorc** [comment]
@fbaligand 

sorry for being unclear.

> But if I understood well, each sbc log has its own call id.

yes, that is true

> So, given that task_id is "call_ids[0]", there is no risk that 2 sbc logs share the same aggregate map. No ?

yes.

however, the second sbc log typically arrives after the mgc log and so if it is dropped it will simply disappear. its the drop i was trying to avoid rather than a shared aggregate map.

**Comment by fbaligand** [comment]
OK, that confirms my mind : this is not useful, and worse, it can led to bug.
Anyway, my first configuration don't solve your problem.
To be honest, I haven't tested it.

**Comment by fbaligand** [comment]
Now, here's a solution which works for all cases (sdc, then mgc, then sdc).
And this time, I tested it :)

``` ruby
filter  {
	
	if ([call_ids][0] != "") {
	
		if [type] == 'sbc' {
			aggregate {
				task_id => "%{call_ids[0]}"
				code => "map.merge!(event.to_hash)"
			}
			drop {}
		}

		if [type] == 'mgc' {
			ruby {
				code => "event.set('call_id_link', event.get('[@metadata][fingerprint]'))"
			}
			clone {
				clones => ["merge_sbc_log"]
			}
		}

		if [type] == "merge_sbc_log" {
			split {
				field => "call_ids"
			}
			ruby {
				code => "event.set('call_ids', [event.get('call_ids')])"
			}
			aggregate {
				task_id => "%{call_ids[0]}"
				code => "map['call_id_link'] = event.get('call_id_link') ; event.cancel"
				push_map_as_event_on_timeout => true
				timeout => 5
				# timeout => 60
			}
			# drop {}
		}
	}
}
```

For information, I tested with following input data, and it works fine :
``` json
{"type" : "sbc", "call_ids" : ["1"]}
{"type" : "mgc", "call_ids" : ["1", "2"], "@metadata" : {"fingerprint" : "fp"} }
{"type" : "sbc", "call_ids" : ["2"]}
```

One important thing : the final sdc events will arrive after a 5s timeout.

**Comment by fbaligand** [comment]
My previous configuration is interesting, but has one disadvantage : sbc logs are released from aggregate maps only after timeout happens.
This works fine, but if you have lots of incoming events at a time, this will led to memory issue, potentially a OutOfMemory error.

So, here a little bit more complex configuration, but that is not based on a timeout mechanism.
I tested it with same input data than in my previous comment, and it works fine.

``` ruby
filter  {
	
	if ([call_ids][0] != "") {
	
		if [type] == 'sbc' {
			aggregate {
				task_id => "%{call_ids[0]}"
				code => "map.merge!(event.to_hash) ; event.cancel()"
				map_action => "create"
			}
			aggregate {
				task_id => "%{call_ids[0]}"
				code => "event.set('call_id_link', map['call_id_link'])"
				map_action => "update"
				end_of_task => true
				timeout => 60
			}
		}

		if [type] == 'mgc' {
			ruby {
				code => "event.set('call_id_link', event.get('[@metadata][fingerprint]'))"
			}
			clone {
				clones => ["merge_sbc_log"]
			}
		}

		if [type] == "merge_sbc_log" {
			split {
				field => "call_ids"
			}
			ruby {
				code => "event.set('call_ids', [event.get('call_ids')])"
			}
			aggregate {
				task_id => "%{call_ids[0]}"
				code => "map['call_id_link'] = event.get('call_id_link') ; event.cancel()"
				map_action => "create"
			}
			aggregate {
				task_id => "%{call_ids[0]}"
				code => "
					map['call_id_link'] = event.get('call_id_link')
					event.overwrite(LogStash::Event.new(map))
				"
				map_action => "update"
				end_of_task => true
			}
		}
	}
}
```

**Comment by crollorc** [comment]
@fbaligand 

Using the initial code:
```
        if ([call_ids][0] != "") {
                if ([type] == "sbc") {
                        aggregate {
                                task_id         => "%{call_ids[0]}"
                                code            => "map.merge!(event.to_hash)"
                        }
                        drop {}
                }
                if ([type] == "mgc") {
                        ruby {
                                code            => "event.set('call_id_link', event.get('[@metadata][fingerprint]'))"
                        }
                        clone {
                                clones          => ["merge_sbc_log"]
                        }
                }
                if ([type] == "merge_sbc_log") {
                        split {
                                field           => "call_ids"
                        }
                        ruby {
                                code            => "event.set('call_ids', [event.get('call_ids')])"
                        }
                        aggregate {
                                task_id         => "%{call_ids[0]}"
                                code            => "map['call_id_link'] = event.get('call_id_link');event.cancel"
                                push_map_as_event_on_timeout => true
                                timeout         => 5
                        }
                }
        }

```
the MGC event gets through but the SBC events have some problems.

The first SBC event is constantly updated due to an invalid id  and only has a valid call_id_link:
```
{
  "_index": "logstash-%{type}-2016.12.20",
  "_type": "logs",
  "_id": "%{[@metadata][fingerprint]}",
  "_score": null,
  "_source": {
    "@timestamp": "2016-12-20T10:30:34.983Z",
    "call_id_link": "cec36824eff2813b2086bac8e1ddb33b91422f20",
    "@version": "1",
    "tags": []
  },
  "fields": {
    "@timestamp": [
      1482229834983
    ]
  },
  "sort": [
    1482229834983
  ]
}

```
the second event is again constantly updated due to an invalid id:
```
{
  "_index": "logstash-sbc-2016.12.20",
  "_type": "sbc",
  "_id": "%{[@metadata][fingerprint]}",
  "_score": null,
  "_source": {
    "release": "F10",
    "duration_inc_ringing_setup": 48,
    "duration_only_setup": 48,
    "type": "sbc",
    "dt": "20161220 102917",
    "duration": 0,
    "path": "/cirpack_cdr/IR_SBC/grnti_201612200000",
    "circut_ids": "X",
    "trunk_outbound": "INT_TPIP_IE",
    "@version": "1",
    "host": "voice-gw",
    "trunk_inbound": "IC_IE_VSIP37387_BT_IE_2",
    "called_number_zone": "north america",
    "calling_number_cc": "China",
    "calling_nature_post": 0,
    "calling_number_post": "X",
    "duration_inc_ringing": 0,
    "app": "I2",
    "called_number_post": "X",
    "calling_number": "8613883200848",
    "called_nature_post": 0,
    "calling_nature": 4,
    "dtx": "20161220 10291783",
    "message": "I2$188 T0b9 0 1 20161220 102917 0 0 48 ac1a820b 0 0 0 000 4 8613883200848@62.172.16.204:5060 0 X 0 0 4 19495051443@84.203.130.151 000 0 0 X 0 0 F10 X X IC_IE_VSIP37387_BT_IE_2 172.26.160.20:5061 0 Z03a 54021041-3691218557-839532@sbc-ire-db-bm02a.btireland.com X028 \"IBCF-2-148222975783995653506141\" X X X G021 20161220 10291783 ? ? 0000004718 O02d 48000 160 0 1 0 0 0 0 X 84.203.130.155:34972",
    "call_ids": [
      "54021041-3691218557-839532"
    ],
    "calling_number_zone": "east asia",
    "tags": [],
    "@timestamp": "2016-12-20T10:29:17.830Z",
    "called_nature": 4,
    "duration_only_ringing": 0,
    "call_id_link": "a30089b71a5012e825ae2eaadb0e0cca066a97e3",
    "called_number": "19495051443",
    "called_number_cc": "United States",
    "device": "ie_sbc"
  },
  "fields": {
    "@timestamp": [
      1482229757830
    ]
  },
  "sort": [
    1482229757830
  ]
}
```

I tried to add a new fingerprint for each cloned/split event but that didn't help:
```
        if ([call_ids][0] != "") {
                if ([type] == "sbc") {
                        aggregate {
                                task_id         => "%{call_ids[0]}"
                                code            => "map.merge!(event.to_hash)"
                        }
                        drop {}
                }
                if ([type] == "mgc") {
                        ruby {
                                code            => "event.set('call_id_link', event.get('[@metadata][fingerprint]'))"
                        }
                        clone {
                                clones          => ["merge_sbc_log"]
                        }
                }
                if ([type] == "merge_sbc_log") {
                        split {
                                field           => "call_ids"
                        }
                        ruby {
                                code            => "event.set('call_ids', [event.get('call_ids')])"
                        }
                        fingerprint {
                                method          => "SHA1"
                                key             => "blahdiblah"
                                target          => "[@metadata][fingerprint]"
                        }
                        aggregate {
                                task_id         => "%{call_ids[0]}"
                                code            => "map['call_id_link'] = event.get('call_id_link');event.cancel"
                                push_map_as_event_on_timeout => true
                                timeout         => 5
                        }
                }
        }

```

**Comment by fbaligand** [comment]
- The first problem is because the sdc event arrives later than 5 seconds after mgc event.
So aggregate timeout has been launched with only call_id_link into agg map.

- The second problem is because, you configured a bad `_id` content.
For sdc logs, `_id` field should contain `%{call_id_link}`

**Comment by fbaligand** [comment]
That said, if you use mgc fingerprint like elasticsearch id for sbc logs, you will have problems...
elasticsearch id must be unique, so if multiple sbc documents have the same id, the second sbc doc will replace the first one in elasticsearch.

**Comment by crollorc** [comment]
@fbaligand 

i made a few changes to try to get the fingerprints working.

1) the fingerprint is now stored in the event rather than in metadata:
```
        # ... generate fingerprint to ensure deduplication
        fingerprint {
                method  => "SHA1"
                key     => "blahdiblah"
        }
```

```
2) i generate a new fingerprint for each of the cloned/split events:
        if ([call_ids][0] != "") {
                if ([type] == "sbc") {
                        aggregate {
                                task_id         => "%{call_ids[0]}"
                                code            => "map.merge!(event.to_hash)"
                        }
                        drop {}
                }
                if ([type] == "mgc") {
                        ruby {
                                code            => "event.set('call_id_link', event.get('fingerprint'))"
                        }
                        clone {
                                clones          => ["merge_sbc_log"]
                        }
                }
                if ([type] == "merge_sbc_log") {
                        split {
                                field           => "call_ids"
                        }
                        ruby {
                                code            => "event.set('call_ids', [event.get('call_ids')])"
                        }
                        fingerprint {
                                method          => "SHA1"
                                key             => "blahdiblah"
                        }
                        aggregate {
                                task_id         => "%{call_ids[0]}"
                                code            => "
                                                        map['call_id_link'] = event.get('call_id_link')
                                                        map['fingerprint'] = event.get('fingerprint')
                                                        event.cancel
                                                "
                                push_map_as_event_on_timeout => true
                                timeout         => 30
                        }
                }
        }

```
3) the first problem i have noticed is that i see extra incomplete events in some cases ~ 75s after mgc+sbc events.

i think this may be because in some cases the mgc log will include 1 or more call_ids for which there is no sbc log.

any ideas on how to address that problem?

**Comment by fbaligand** [comment]
Concerning point 2, now, I don't understand your need. 
If you generate a fingerprint per clone, it is not any more a shared field between sdc and mgc logs. And potentially, you can directly generate fingerprint in sdc event, without using aggregate plugin. 
Then, what fingerprint is for ? 
If it is intended to generate a unique elasticsearch id, you are not obliged to do that. By default, elasticsearch will generate itself a unique id for each indexed document.

**Comment by crollorc** [comment]
@fbaligand 

Apologies again for causing confusion.

> Then, what fingerprint is for ?

I use fingerprints so that if logfiles need to be reloaded they will result in event updates rather than new events.

The call_id_link on the other hand is there to link logs which relate to the same telephone call.

**Comment by fbaligand** [comment]
OK. Thanks for the explanation.
So in mdc logs, you finally need both a own fingerprint and a call_id_link, which is mgc log fingerprint. Right ?

**Comment by crollorc** [comment]
> So in mdc logs, you finally need both a own fingerprint and a call_id_link, which is mgc log fingerprint. Right ?

Yes and when choosing a value for the call_id_log, I used the fingerprint simply because I knew it would be unique.

**Comment by crollorc** [comment]
@fbaligand

I've added the second version of the code you listed above now and it appears to be working by and large which is great!
```
        if ([call_ids][0] != "") {
                if ([type] == "sbc") {
                        aggregate {
                                task_id         => "%{call_ids[0]}"
                                code            => "map.merge!(event.to_hash);event.cancel()"
                                map_action      => "create"
                        }
                        aggregate {
                                task_id         => "%{call_ids[0]}"
                                code            => "event.set('call_id_link', map['call_id_link'])"
                                map_action      => "update"
                                end_of_task     => true
                                timeout         => 60
                        }
                }
                if ([type] == "mgc") {
                        ruby {
                                code            => "event.set('call_id_link', event.get('fingerprint'))"
                        }
                        clone {
                                clones          => ["merge_sbc_log"]
                        }
                }
                if ([type] == "merge_sbc_log") {
                        split {
                                field           => "call_ids"
                        }
                        ruby {
                                code            => "event.set('call_ids', [event.get('call_ids')])"
                        }
                        fingerprint {
                                method          => "SHA1"
                                key             => "blahdiblah"
                        }
                        aggregate {
                                task_id         => "%{call_ids[0]}"
                                code            => "
                                                        map['call_ids'] = event.get('call_ids')
                                                        map['call_id_link'] = event.get('call_id_link')
                                                        map['fingerprint'] = event.get('fingerprint')
                                                        event.cancel
                                                "
                                map_action      => "create"
                        }
                        aggregate {
                                task_id         => "%{call_ids[0]}"
                                code            => "
                                                        map['call_id_link'] = event.get('call_id_link')
                                                        event.overwrite(LogStash::Event.new(map))
                                                "
                                map_action      => "update"
                                end_of_task     => true
                        }
                }
        }
```

However, this has highlighted a few logic problems that I was not aware of and I'm unsure how to deal with.

By looking at empty events which are coming when I guess their aggregation timer expires, I can see that:
1) In some cases >1 MGC logs will share a call_id
2) In some cases >1 SBC logs will share a call_id

Do you have any ideas on how to incorporate those events?

Thanks again for all your time.

**Comment by fbaligand** [comment]
2 things about fingerprint :
- if you don't define `sources` option, by default, only `message` field is used to build fingerprint.
So I advice you to define `sources` option with all fields that are relevant to build fingerprint
- in fingerprint plugin, you can define the `target` field where to store fingerprint. Using `target` option.
That is more clean than copy after the fingerprint in another field using `ruby` filter.

**Comment by crollorc** [comment]
thanks for the advise on fingerprint @fbaligand and apologies for using up your time

**Comment by fbaligand** [comment]
Concerning your last posted  configuration :
- nice to see that we have something that starts to answer your need !
- using this second solution, when you have a mgc log which contains call_ids with no associated sdc log, this will create a aggregate map for this call id, that will expire in 60s and so be dropped automatically by aggregate plugin
- Then, it is useless to define call_ids and fingerprint fields in aggregate map, given that it is not used :
```
code            => "
                                                        map['call_ids'] = event.get('call_ids')
                                                        map['call_id_link'] = event.get('call_id_link')
                                                        map['fingerprint'] = event.get('fingerprint')
                                                        event.cancel
                                                "
```
- I advice you to generate fingerprint for sdc logs just before put sdc logs into an aggregate map. It should work better than generate fingerprint on mgc clones... that are finally not used in the last config you posted.

So the final config could be :
```
if ([call_ids][0] != "") {
                if ([type] == "sbc") {
                        fingerprint {
                                method          => "SHA1"
                                key             => "blahdiblah"
                               sources => ["field1", "field2", ...] # TO BE DEFINED
                        }
                        aggregate {
                                task_id         => "%{call_ids[0]}"
                                code            => "map.merge!(event.to_hash);event.cancel()"
                                map_action      => "create"
                        }
                        aggregate {
                                task_id         => "%{call_ids[0]}"
                                code            => "event.set('call_id_link', map['call_id_link'])"
                                map_action      => "update"
                                end_of_task     => true
                                timeout         => 60
                        }
                }
                if ([type] == "mgc") {
                        ruby {
                                code            => "event.set('call_id_link', event.get('fingerprint'))"
                        }
                        clone {
                                clones          => ["merge_sbc_log"]
                        }
                }
                if ([type] == "merge_sbc_log") {
                        split {
                                field           => "call_ids"
                        }
                        ruby {
                                code            => "event.set('call_ids', [event.get('call_ids')])"
                        }
                        aggregate {
                                task_id         => "%{call_ids[0]}"
                                code            => "
                                                        map['call_id_link'] = event.get('call_id_link')
                                                        event.cancel
                                                "
                                map_action      => "create"
                        }
                        aggregate {
                                task_id         => "%{call_ids[0]}"
                                code            => "
                                                        map['call_id_link'] = event.get('call_id_link')
                                                        event.overwrite(LogStash::Event.new(map))
                                                "
                                map_action      => "update"
                                end_of_task     => true
                        }
                }
        }
```

**Comment by fbaligand** [comment]
> By looking at empty events which are coming when I guess their aggregation timer expires, I can see that:
>    In some cases >1 MGC logs will share a call_id
>    In some cases >1 SBC logs will share a call_id
>Do you have any ideas on how to incorporate those events?

If you have such cases, that's really pesky...
- this is a problem for aggregate task_id mechanism, which is based on the fact `task_id` is an id.
- this is potentially a problem for fingerprint, if it led to 2 different events that generate the same fingerprint.

That said, the main question is : what do you want to do with duplicate mgc logs and sbc logs ?

**Comment by crollorc** [comment]
@fbaligand 

> That said, the main question is : what do you want to do with duplicate mgc logs and sbc logs ?

Unfortunately, the events are not simply duplicates. 

Instead they are edge cases i was not aware of where a call_id appears in multiple MGC logs and or multiple SBC logs.

Ideally we would like all these events to include the same call_id_link.

So, I guess an example would be:
	1) SBC log #1

```
 	log: I2$190 T0b4 0 1 20161214 082233 4 4 11 ac1a800b 0 0 0 000 4 390694800319 0 X 0 0 4 393484959705 000 0 0 X 0 0 F10 X X IC_UK_VSIP38362_OMEGA_1 172.26.160.20:5060 0 Z035 7c58b6906383931c66a7d8d2530c6b3c X028 "IBCF-2-148170375385588622905679" X X X G02a 20161214 08223385 ? 0000000710 0000001080 O036 11000 60800 16640 380 104 0 0 0 X 212.126.62.75:34156
 	call_ids: 7c58b6906383931c66a7d8d2530c6b3c
	call_id_link: 63420e94e87d9a17f65162e83e948468cd419c52
```

	2) MGC log #1

```
	log: RP$25d T0db 1 0 20161214 082234 4 9 10 ac1a8015 0 0 1 00A 4 390694800319 4 390694800319 0 1 4 393484959705 001 0 4 393484959705 1 8 F10 FROM_GCS_UK2 314702-249169 loop[IC_UK_VSIP38362_OMEGA_1] loop[gw_sip,IC_UK_VSIP37927_BICS_2] 0 D00d 14894 50f0c6 E007 020100 H01b 4 390694800319 10 SERVNODE S015 1 1 4 393484959705 0 G033 20161214 08223402 0000000183 0000000693 0000001065 V013 8 20 0 0 0 SIP_OUT V012 8 20 0 0 0 SIP_IN Z02e 27047-VV-06dfc1f0-56ed86442 V013 8 20 0 0 0 SIP_OUT Z02e 28939-FV-06dfc1ee-3fe84a435 Z035 7c58b6906383931c66a7d8d2530c6b3c
	call_ids: 27047-VV-06dfc1f0-56ed86442, 28939-FV-06dfc1ee-3fe84a435, 7c58b6906383931c66a7d8d2530c6b3c
	call_id_link: 63420e94e87d9a17f65162e83e948468cd419c52
```

	3) SBC log #2

```
	log: I2$1ad T0d8 0 0 20161214 082234 4 4 10 ac1a800b 0 0 0 000 4 390694800319 4 390694800319 0 0 4 393484959705 000 0 0 X 0 0 F10 X X 172.26.160.20:5060 IC_UK_VSIP37927_BICS_2 0 Z02e 27047-VV-06dfc1f0-56ed86442 X028 "IBCF-2-148170375412027222905684" X X X G02a 20161214 08223412 ? 0000000683 0000001056 O036 10000 16640 61120 104 382 0 0 0 X 212.126.62.77:34864
	call_ids: 27047-VV-06dfc1f0-56ed86442
	call_id_link: 63420e94e87d9a17f65162e83e948468cd419c52
```

	4) MGC log #2

```
	log: GI$214 T0ae 38142 1 20161220 150223 95 97 98 ac1a8015 1 0 1 00A 4 35316582000 4 35314199521 0 1 4 35314051178 001 0 4 35314051178 1 24 B10 XX 302554-214463 loop SIP[sip.ie.viatel.com] 0 D00d 14894 c02133 G033 20161220 15022386 0000000074 0000000234 0000013722 E007 020100 Y008 118 118 V012 8 20 0 0 0 SIP_IN V013 8 20 0 0 0 SIP_OUT O037 1518 12300 12300 75 75 0 0 0 IP2TDM 172.26.162.5:31712 V013 8 20 0 0 0 SIP_OUT V013 8 20 0 0 0 SIP_OUT Z02e 27047-VV-06dfc1f0-56ed86442
	call_ids: 27047-VV-06dfc1f0-56ed86442
	call_id_link: 63420e94e87d9a17f65162e83e948468cd419c52
```

	5) SBC log #3

```
	log: IB$154 T0c4 0 0 20161220 150223 0 0 1 ac1a820b 0 0 0 000 0 016582000 0 014199521 0 0 0 0035314051178 000 0 0 X 0 0 B1a6 X X 172.26.160.70:5071 IE_CORE_IFJ 0 Z02e 7c58b6906383931c66a7d8d2530c6b3c X027 "IBCF-0-14822461439620559577330" X X X G021 20161220 15022396 ? ? 0000000021
	call_ids: 7c58b6906383931c66a7d8d2530c6b3c
	call_id_link: 63420e94e87d9a17f65162e83e948468cd419c52
```

**Comment by fbaligand** [comment]
- To have the same call_id_link, the important thing is what fields you set in `source` fingerprint option.
- That said, concerning aggregate plugin usage for your need, this becomes a lot more complicated...
  - With several sbc logs with same call id, you have to find a composite key for `task_id` that is really unique for each sbc log. Something like `%{call_ids[0]}%{another_id_field}`. And worse, if you want to share call_id_link from mgc logs to sbc logs, this `another_id_field` should be known in mgc logs...
  - With several mgc logs, you'll have to process the case where you have first mgc log, then second mgc log (before the sbc log). You can do that, by replacing the last aggregate filter by the following code  :
``` ruby
aggregate {
                                task_id         => "%{call_ids[0]}"
                                code            => "
                                                        map['call_id_link'] = event.get('call_id_link')
                                                        event.overwrite(LogStash::Event.new(map)) if map.has_key?('type')
                                                "
                                map_action      => "update"
                                end_of_task     => true
}
if [type] != "sbc" {
    aggregate {
                                task_id         => "%{call_ids[0]}"
                                code            => "
                                                        map['call_id_link'] = event.get('call_id_link')
                                                        event.cancel
                                                "
                                map_action      => "create"
                        }
    }
```

For the first aggregate point... good luck :)

**Comment by crollorc** [comment]
@fbaligand 

> With several mgc logs, you'll have to process the case where you have first mgc log, then second mgc log (before the sbc log). You can do that, by replacing the last aggregate filter by the following code :

I found an identifier which can be used to link all related MGC logs which should make things simpler. It is called 'ref and looks like this `cc9d51`. I guess this could be used in an additional aggregate to apply the shared call_id_link to those events.

If so, should I add this aggregate like the below?

```
                if ([type] == "mgc") {
                        ruby {
                                code            => "event.set('call_id_link', event.get('fingerprint'))"
                        }
			if ([ref]) {
                               aggregate {
                                       task_id         => "%{ref}"
                                       code            => "map.merge!(event.to_hash);event.cancel()"
                                       map_action      => "create"
                               }
                               aggregate {
                                       task_id         => "%{ref}"
                                       code            => "event.set('call_id_link', map['call_id_link'])"
                                       map_action      => "update"
                                       end_of_task     => true
                                       timeout         => 60
                               }
			}
```

> With several sbc logs with same call id, you have to find a composite key for task_id that is really unique for each sbc log. Something like %{call_ids[0]}%{another_id_field}. And worse, if you want to share call_id_link from mgc logs to sbc logs, this another_id_field should be known in mgc logs...

Yikes! Where we have the same call_id in multiple SBC logs, is there a way of at least ensuring the events make it through logstash as many of these events appear to be dropped/cancelled at present.

Thanks again for your assistance Fabien.

**Comment by fbaligand** [comment]
Concerning your sample aggregate configuration, I'm not sure what you are trying to do.
With the code you show, you will merge 2 mgc logs into one. So at end, only one elasticsearch doc for 2 original mgc logs. Is that what you want to do ?

**Comment by crollorc** [comment]
@fbaligand 

> With the code you show, you will merge 2 mgc logs into one. So at end, only one elasticsearch doc for 2 original mgc logs. Is that what you want to do ?

No, sorry. I would like 2 MGC logs with a shared call_id_link. I think this is working now with the following snippet before the `if ([call_ids][0] != "") {` statement:
````
        if ([ref]) {
                aggregate {
                        task_id                 => "%{ref}"
                        code                    => "
                                                        map['call_id_link'] ||= event.get('fingerprint')
                                                        event.set('call_id_link', map['call_id_link'])
                                                "
                        timeout                 => 60
                }
        }
````

Does that look OK?

Also, where we have the same call_id in multiple SBC logs, is there a way of at least ensuring the events make it through logstash as many of these events appear to be dropped/cancelled at present?

Finally, where we have 2 MGC logs, the second log is tagged with type 'merge_sbc_log' which is incorrect and results in it being stored to the wrong index.

**Comment by crollorc** [comment]
> Finally, where we have 2 MGC logs, the second log is tagged with type 'merge_sbc_log' which is incorrect and results in it being stored to the wrong index.

Actually, it looks like we have a duplicate of the first MGC log with a type of 'merge_sbc_log'.

**Comment by fbaligand** [comment]
Concerning sharing the `call_id_link` between mgc logs, the right code should be :
``` ruby
if ([ref]) {
                fingerprint {
                       ... 
                       target => "call_id_link"
                 }
                aggregate {
                        task_id                 => "%{ref}"
                        code                    => "map['call_id_link'] = event.get('call_id_link')"
                        map_action => "create"
                }
                aggregate {
                        task_id                 => "%{ref}"
                        code                    => "event.set('call_id_link', map['call_id_link'])"
                        map_action => "update"
                        end_of_task => true
                        timeout                 => 60
                }
        }
```

If there is only one mgc, log for `ref`, then the aggregate map will expire in 60s and be deleted.

**Comment by fbaligand** [comment]
To avoid to lost some sbc logs, here's the global configuration.
The 2 new parts are :
- the 2th aggregate filter has a `push_map_as_event_on_timeout => true` with a timeout tag.
- at the end, we test `timeout` events, and remove ones which are not sbc (these are mgc clones)

``` ruby
	if ([call_ids][0] != "") {
			if ([type] == "sbc") {
					fingerprint {
							method          => "SHA1"
							key             => "blahdiblah"
						   sources => ["message"] # TO BE CUSTOMIZED ?
						   target => "fingerprint"
					}
					aggregate {
							task_id         => "%{call_ids[0]}"
							code            => "map.merge!(event.to_hash);event.cancel()"
							map_action      => "create"
					}
					aggregate {
							task_id         => "%{call_ids[0]}"
							code            => "event.set('call_id_link', map['call_id_link'])"
							map_action      => "update"
							end_of_task     => true
							push_map_as_event_on_timeout => true
							timeout         => 60
							timeout_tags => ["timeout"]
					}
			}
			if ([type] == "mgc") {
					ruby {
							code            => "event.set('call_id_link', event.get('fingerprint'))"
					}
					clone {
							clones          => ["merge_sbc_log"]
					}
			}
			if ([type] == "merge_sbc_log") {
					split {
							field           => "call_ids"
					}
					ruby {
							code            => "event.set('call_ids', [event.get('call_ids')])"
					}
					aggregate {
							task_id         => "%{call_ids[0]}"
							code            => "
													map['call_id_link'] = event.get('call_id_link')
													event.cancel
											"
							map_action      => "create"
					}
					aggregate {
							task_id         => "%{call_ids[0]}"
							code            => "
													map['call_id_link'] = event.get('call_id_link')
													event.overwrite(LogStash::Event.new(map))
											"
							map_action      => "update"
							end_of_task     => true
					}
			}
	}
	
	if "timeout" in [tags] {
		if [type] != "sbc" {
			drop{}
		}
	}
```

**Comment by fbaligand** [comment]
I just fixed the code above.

**Comment by crollorc** [comment]
the first problem is that where we have a two mgc logs and no sbc logs or call_ids, different `call_id_link`s are added. i guess this is related to this code snippet from earlier?
```
if ([ref]) {
                fingerprint {
                       ... 
                       target => "call_id_link"
                 }
                aggregate {
                        task_id                 => "%{ref}"
                        code                    => "map['call_id_link'] = event.get('call_id_link')"
                        map_action => "create"
                }
                aggregate {
                        task_id                 => "%{ref}"
                        code                    => "event.set('call_id_link', map['call_id_link'])"
                        map_action => "update"
                        end_of_task => true
                        timeout                 => 60
                }
        }
```

**Comment by fbaligand** [comment]
Question : if `ref` is the correlation id between mgc logs, why not just use `ref` as `call_id_link` to share with sdc logs ?

**Comment by fbaligand** [comment]
Anyway, my configuration above was wrong. I have not tested it :)
Here's the right configuration, which is tested :
``` ruby 
					aggregate {
							task_id                 => "%{ref}"
							code                    => "map['call_id_link'] = event.get('call_id_link')"
							map_action => "create"
							add_tag => ["first_one"]
					}
					if "first_one" not in [tags] {
						aggregate {
								task_id                 => "%{ref}"
								code                    => "event.set('call_id_link', map['call_id_link'])"
								map_action => "update"
								end_of_task => true
								timeout                 => 60
						}
					}
```

**Comment by fbaligand** [comment]
A remark that could be important : 
could you have maximum 2 mgc logs sharing call_id ? Or more ?

**Comment by crollorc** [comment]
> Question : if ref is the correlation id between mgc logs, why not just use ref as call_id_link to share with sdc logs ?

Yes, we could use the ref as the link between the mgc and sbc logs.

> could you have maximum 2 mgc logs sharing call_id ? Or more ?

Its possible there could be more than two in some edge cases.

> Anyway, my configuration above was wrong. I have not tested it :)

Thank you. I'll test that out.

**Comment by fbaligand** [comment]
If you use `ref` as `call_id_link`, you don't have to create a fingerprint, and then use aggregate filters to share it between mgc logs, given that they already have `ref` field.
So you greatly simplify your logstash configuration.

That said, if you still want to generate call_id_link and share it between mgc logs with same `ref`, given that there could have more than 2 mgc logs with same `ref`, you have to remove line `end_of_task => true` in my last configuration suggestion.
Doing that, aggregate map will stay during 60s for all mgc logs with same `ref` that arrive and then be dropped.

**Comment by crollorc** [comment]
> That said, if you still want to generate call_id_link and share it between mgc logs with same ref, given that there could have more than 2 mgc logs with same ref, you have to remove line end_of_task => true in my last configuration suggestion.

I see no problems with the MGC logs so far after making the changes you  suggested.

> If you use ref as call_id_link, you don't have to create a fingerprint, and then use aggregate filters to share it between mgc logs, given that they already have ref field.
> So you greatly simplify your logstash configuration.

I've made this change now and it makes the config a lot cleaner.

Currently the only outstanding issue is the problem of SBC logs sharing the same call_id and cases I've found so far as at least the events are not being dropped.

Thanks again for all your help Fabien.

**Comment by fbaligand** [comment]
Globally, I think your use case wins the international award of the most complete aggregate plugin usage !🥇

**Comment by fbaligand** [comment]
For SBC logs, so if I understand well, with my last configuration suggestion, you loose not anymore any SBC log ? Right ?

Now, I will work on a solution that set `call_id_link` on every sbc log, whatever the case.
By the way, did you keep `call_id_link` or `ref` like shared field name ?

**Comment by crollorc** [comment]
> For SBC logs, so if I understand well, with my last configuration suggestion, you loose not anymore any SBC log ? Right ?

yes.

> By the way, did you keep call_id_link or ref like shared field name ?

i'm now using `ref` and, as you suggested, it makes the config quite a bit simpler.

**Comment by fbaligand** [comment]
OK, so I finally found a configuration which allows not only to not loose any sbc event, but also that every sbc event has its `ref` field filled !

The configuration is quite different from before, with less aggregate filters, but more code inside.
I looked at optimizing memory usage, by freeing sbc events as soon as possible.

I tested this configuration with all possibilities, and up to me, it works for all cases !

``` ruby
	if ([call_ids][0] != "") {
			if ([type] == "sbc") {
					fingerprint {
							method          => "SHA1"
							key             => "blahdiblah"
						   source => ["message"]
						   target => "fingerprint"
					}
					aggregate {
							task_id         => "%{call_ids[0]}"
							code            => "
													if map.has_key?('ref')
														event.set('ref', map['ref']) 
													else
														map['sbc_logs'] ||= []
														map['sbc_logs'] << event.to_hash
														event.cancel()
													end
											   "
					}
			}
			
			if ([type] == "mgc") {
					clone {
							clones          => ["merge_sbc_log"]
					}
			}
			
			if ([type] == "merge_sbc_log") {
					split {
							field           => "call_ids"
					}
					ruby {
							code            => "event.set('call_ids', [event.get('call_ids')])"
					}
					aggregate {
							task_id         => "%{call_ids[0]}"
							code            => "
													map['ref'] ||= event.get('ref')
													if map.has_key?('sbc_logs')
														map['sbc_logs'].each do |sbc_log|
															sbc_log['ref'] = map['ref']
														end
														event.set('sbc_logs', map['sbc_logs'])
														map['sbc_logs'].clear()
													else
														event.cancel()
													end
											   "
							timeout         => 60
					}
					split {
							field           => "sbc_logs"
					}
					ruby {
						code => "event.overwrite(LogStash::Event.new(event.get('sbc_logs')))"
					}
			}
	}
```

**Comment by crollorc** [comment]
> OK, so I finally found a configuration which allows not only to not loose any sbc event, but also that every sbc event has its ref field filled !

Wow. That config looks like it is working, thanks Fabien!

Occasionally we will have buggy SBC logs which are generated at the start of the call while the other SBC and MGC logs are generated at the end of the call.

Where a call lasts longer than the aggregation timeout will these buggy SBC logs be dropped or passed through? If passed through could we tag these so they can be easily found?

**Comment by fbaligand** [comment]
I work a little more this morning to process the case where a sbc log has no related mgc log.
Here's the configuration. The main update is `push_map_as_event_on_timeout => true`

``` ruby
	if ([call_ids][0] != "") {
			if ([type] == "sbc") {
					fingerprint {
							method          => "SHA1"
							key             => "blahdiblah"
						   source => ["message"]
						   target => "fingerprint"
					}
					aggregate {
							task_id         => "%{call_ids[0]}"
							code            => "
													if map.has_key?('ref')
														event.set('ref', map['ref']) 
													else
														map['sbc_logs'] ||= []
														map['sbc_logs'] << event.to_hash
														event.cancel()
													end
											   "
					}
			}
			
			if ([type] == "mgc") {
					clone {
							clones          => ["merge_sbc_log"]
					}
			}
			
			if ([type] == "merge_sbc_log") {
					split {
						field => "call_ids"
					}
					ruby {
						code => "event.set('call_ids', [event.get('call_ids')])"
					}
					aggregate {
						task_id => "%{call_ids[0]}"
						code => "
							map['ref'] ||= event.get('ref')
							if map.has_key?('sbc_logs')
								event.set('sbc_logs', map.delete('sbc_logs'))
							else
								event.cancel()
							end
						"
						timeout         => 60
						push_map_as_event_on_timeout => true
					}
					if ![sbc_logs] {
						drop {}
					}
					split {
						field => "sbc_logs"
					}
					if [ref] {
						mutate {
							add_field => { "[sbc_logs][ref]" => "%{ref}" }
						}
					}
					ruby {
						code => "event.overwrite(LogStash::Event.new(event.get('sbc_logs')))"
					}
			}
	}
```

**Comment by fbaligand** [comment]
> Where a call lasts longer than the aggregation timeout will these buggy SBC logs be dropped or passed through? If passed through could we tag these so they can be easily found?

I just updated my previous configuration to add a timeout tag in that case.

Anyway, choose carefully your timeout value :
- too short : you loose ref field in sbc logs
- too long : you risk logstash big memory consuming, and timeout sbc logs arrive lately in elasticsearch.

**Comment by crollorc** [comment]
> I just updated my previous configuration to add a timeout tag in that case.

That is working for me, thank you.

> Where a call lasts longer than the aggregation timeout will these buggy SBC logs be dropped or passed through? If passed through could we tag these so they can be easily found?

One additional thing in relation to this. If the call is shorter that the aggregation timer, then the SBC and MGC logs will be linked by `ref`. However, the call `duration`s will be mismatched as the buggy SBC log always records `duration` as zero. Where could I add check to add a tag like `buggy_sbc` where this duration mismatch occurs?

**Comment by fbaligand** [comment]
Duration... a new data which we didn't talk before :)

How this duration is computed?

**Comment by crollorc** [comment]
> How this duration is computed?

No computation needed as each of the logs includes the call duration in seconds and this is groked from the logs into `duration`!

**Comment by fbaligand** [comment]
Ok. 
It is in mgc or sbc logs ?

**Comment by crollorc** [comment]
> It is in mgc or sbc logs ?

All logs have a duration field.

However with our buggy SBC logs, the SBC log duration would always be 0s even when related MGC and SBC logs have a duration>0s.

**Comment by fbaligand** [comment]
Ok. 
So you want to tag sdc events which have duration = 0 ?

Fabien

**Comment by crollorc** [comment]
> So you want to tag sdc events which have duration = 0 ?

Nearly. I want to tag SBC events with duration = 0 where the linked MGC event has a duration != 0.

**Comment by fbaligand** [comment]
Here's the updated configuration that does that.
For information, it keeps `mgc_duration` field in sdc logs.
If you don't want to keep it, you can use `mutate` filter with `remove_field` option.

``` ruby
	if ([call_ids][0] != "") {
			if ([type] == "sbc") {
					fingerprint {
							method          => "SHA1"
							key             => "blahdiblah"
						   source => ["message"]
						   target => "fingerprint"
					}
					aggregate {
							task_id         => "%{call_ids[0]}"
							code            => "
													if map.has_key?('ref')
														event.set('ref', map['ref'])
														event.set('mgc_duration', map['duration'])
													else
														map['sbc_logs'] ||= []
														map['sbc_logs'] << event.to_hash
														event.cancel()
													end
											   "
					}
			}
			
			if ([type] == "mgc") {
					clone {
							clones          => ["merge_sbc_log"]
					}
			}
			
			if ([type] == "merge_sbc_log") {
					split {
						field => "call_ids"
					}
					ruby {
						code => "event.set('call_ids', [event.get('call_ids')])"
					}
					aggregate {
						task_id => "%{call_ids[0]}"
						code => "
							map['ref'] ||= event.get('ref')
							map['duration'] ||= event.get('duration')
							if map.has_key?('sbc_logs')
								event.set('sbc_logs', map.delete('sbc_logs'))
							else
								event.cancel()
							end
						"
						timeout         => 60
						push_map_as_event_on_timeout => true
					}
					if ![sbc_logs] {
						drop {}
					}
					split {
						field => "sbc_logs"
					}
					if [ref] {
						mutate {
							add_field => {
								"[sbc_logs][ref]" => "%{ref}"
								"[sbc_logs][mgc_duration]" => "%{duration}"
							}
						}
					}
					else {
						mutate {
							add_field => { "[sbc_logs][tags]" => ["timeout"] }
						}
					}
					ruby {
						code => "event.overwrite(LogStash::Event.new(event.get('sbc_logs')))"
					}
			}
			
			if ([type] == "sbc" and [duration] == 0 and [mgc_duration] != 0) {
				mutate {
					add_tag => ["weird_duration"]
				}
			}
	}
```

**Comment by crollorc** [comment]
> Here's the updated configuration that does that.

The tag isn't being added in my case. 

Could it be because the 0s duration SBC log is processed before the >0s duration MGC logs?

**Comment by fbaligand** [comment]
Is your duration field is converted to an integer type in logstash before my configuration ? 

Fabien

**Comment by crollorc** [comment]
> Is your duration field is converted to an integer type in logstash before my configuration ?

It is converted after your config. 

Should I move the conversion before your config?

**Comment by fbaligand** [comment]
Yes. It is required, given that I make integer comparison.

By the way, you could directly convert a field as an INT in grok using that :
`%{INT:duration:int}`

**Comment by crollorc** [comment]
Do I have to convert `mgc_duration` to an int also for this to work? and if so, where?
```
                if ([type] == "sbc" and [duration] == 0 and [mgc_duration] != 0) {
                        mutate {
                                add_tag         => ["duration_mismatch"]
                        }
                }
```

I ask as I am now seeing many SBC logs tagged with a `duration_mismatch` tag where `duration` == `mgc_duration` == 0.

**Comment by fbaligand** [comment]
I invite you to make all parsing/conversion process before  aggregation process. 

Fabien

**Comment by crollorc** [comment]
> I invite you to make all parsing/conversion process before  aggregation process.

Ok, I moved all parsing/conversion and the following above aggregation but I'm still seeing the same issue:
```
        mutate {
                convert                                 => {
                        "mgc_duration"             => "integer"
                }
        }

```

**Comment by fbaligand** [comment]
mgc_duration is not a field from original mgc logs. 
Its original name is duration, in mgc log parsing 

Fabien

**Comment by crollorc** [comment]
> Its original name is duration, in mgc log parsing

Apologies. I've removed the conversion but I still have the same problem with this code:
```
                if ([type] == "sbc" and [duration] == 0 and [mgc_duration] != 0) {
                        mutate {
                                add_tag         => ["duration_mismatch"]
                        }
                }

```
where the tag `duration_mismatch` is being added to SBC events where both `duration=0` and `mgc_duration=0`.

**Comment by fbaligand** [comment]
You have to make duration conversion to integer for mgc logs AND sbc logs. 
Can you check you do that for both ?

Fabien

**Comment by crollorc** [comment]
> Can you check you do that for both ?

the duration conversion is applied to all logs but just to be sure, i added this directly above the aggregation config:

```
        mutate {
                convert                                 => {
                        "duration"                      => "integer"
                }
        }

```

the cases i've found where `mgc_duration > 0` and the related sbc `duration == 0` appear to be correctly tagged with `duration_mismatch`.

and yet, i also see many events where the `mgc_duration == 0` the related sbc `duration == 0` that are also tagged with `duration_mismatch`.

**Comment by fbaligand** [comment]
Honestly, that is really surprising. 
The if statement is a simple Boolean expression. 
There's no reason it works for one case, not for the other. 

That said, maybe you can test with :
[mgc_duration] > 0

**Comment by crollorc** [comment]
> [mgc_duration] > 0

using this:
>                 if ([type] == "sbc" and [duration] == 0 and [mgc_duration] > 0) {
results in some errors like:

```
[2016-12-22T22:05:24,026][ERROR][logstash.pipeline        ] Exception in pipelineworker, the pipeline stopped processing new events, please check your filter configuration and restart Logstash. {"exception"=>#<NoMethodError: undefined method `>' for nil:NilClass>, "backtrace"=>["(eval):4402:in `initialize'", "org/jruby/RubyArray.java:1613:in `each'", "(eval):4400:in `initialize'", "org/jruby/RubyProc.java:281:in `call'", "(eval):4420:in `initialize'", "org/jruby/RubyArray.java:1613:in `each'", "(eval):4414:in `initialize'", "org/jruby/RubyProc.java:281:in `call'", "(eval):1204:in `filter_func'", "/usr/share/logstash/logstash-core/lib/logstash/pipeline.rb:295:in `filter_batch'", "org/jruby/RubyProc.java:281:in `call'", "/usr/share/logstash/logstash-core/lib/logstash/util/wrapped_synchronous_queue.rb:192:in `each'", "org/jruby/RubyHash.java:1342:in `each'", "/usr/share/logstash/logstash-core/lib/logstash/util/wrapped_synchronous_queue.rb:191:in `each'", "/usr/share/logstash/logstash-core/lib/logstash/pipeline.rb:294:in `filter_batch'", "/usr/share/logstash/logstash-core/lib/logstash/pipeline.rb:282:in `worker_loop'", "/usr/share/logstash/logstash-core/lib/logstash/pipeline.rb:258:in `start_workers'"]}
```

**Comment by fbaligand** [comment]
Ok. It's because sometimes mgc_duration is null. 

So try this :
if ([type] == "sbc" and [duration] == 0 and [mgc_duration] and [mgc_duration] > 0) {

**Comment by crollorc** [comment]
> if ([type] == "sbc" and [duration] == 0 and [mgc_duration] and [mgc_duration] > 0) {

ok, unfortunately this results in:
```
[2016-12-22T22:22:05,684][ERROR][logstash.pipeline        ] Exception in pipelineworker, the pipeline stopped processing new events, please check your filter configuration and restart Logstash. {"exception"=>#<ArgumentError: comparison of String with 0 failed>, "backtrace"=>["org/jruby/RubyComparable.java:143:in `>'", "org/jruby/RubyString.java:1865:in `>'", "(eval):4402:in `initialize'", "org/jruby/RubyArray.java:1613:in `each'", "(eval):4400:in `initialize'", "org/jruby/RubyProc.java:281:in `call'", "(eval):4420:in `initialize'", "org/jruby/RubyArray.java:1613:in `each'", "(eval):4414:in `initialize'", "org/jruby/RubyProc.java:281:in `call'", "(eval):1204:in `filter_func'", "/usr/share/logstash/logstash-core/lib/logstash/pipeline.rb:295:in `filter_batch'", "org/jruby/RubyProc.java:281:in `call'", "/usr/share/logstash/logstash-core/lib/logstash/util/wrapped_synchronous_queue.rb:192:in `each'", "org/jruby/RubyHash.java:1342:in `each'", "/usr/share/logstash/logstash-core/lib/logstash/util/wrapped_synchronous_queue.rb:191:in `each'", "/usr/share/logstash/logstash-core/lib/logstash/pipeline.rb:294:in `filter_batch'", "/usr/share/logstash/logstash-core/lib/logstash/pipeline.rb:282:in `worker_loop'", "/usr/share/logstash/logstash-core/lib/logstash/pipeline.rb:258:in `start_workers'"]}
`
```

**Comment by fbaligand** [comment]
That is interesting: it clearly says that mgc_duration is a string here. 

So either you check that duration is really always converted to integer, either just before this if statement, you convert mgc_duration to integer.

**Comment by crollorc** [comment]
> So either you check that duration is really always converted to integer, either just before this if statement, you convert mgc_duration to integer.

i added a conversion just before and the problems have gone away.

**Comment by fbaligand** [comment]
Nice news !
So all your problems are solved ?

**Comment by crollorc** [comment]
> So all your problems are solved ?

Yes but more testing will have to wait until the new year.

Thank you so much for all your help Fabien.

Joyeux Noël et Bonne Année.

**Comment by fbaligand** [comment]
Thank you ! 
I wish you too a merry Christmas and a happy new year !

And... are you French ? :)

**Comment by crollorc** [comment]
> are you French ? :)

Not quite, I'm Irish!

**Comment by fbaligand** [comment]
Okay !
So thanks for the French message ! :)

**Comment by fbaligand** [comment]
Hi @crollorc ,

Have you finished your tests to validate the final configuration ?

**Comment by fbaligand** [comment]
@crollorc 
I close the issue.
